Transformation on a driver level

Mirek0206 opened the pull request.

Pavlucek committed:
Creating decorator TransformingDriver for DriverManager

Pavlucek committed:
Adapt DriverManager to new decorator

Mirek0206 committed:
Revert unnecessary changes.

Pavlucek committed:
Added driver transformations in the TestJobs2dApp.

Mirek0206 committed:
Added moving options.

Mirek0206 committed:
Refactoring.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -63,12 +64,16 @@ private static void setupDrivers(Application application) {
-        driver = new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special");
-        DriverFeature.addDriver("Special line Simulator", driver);
+        // TransformationDriver for Line Simulator
+        TransformingDriver commonTransformedDriver = new TransformingDriver(new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic"));
+        DriverFeature.addDriver("Line Simulator", commonTransformedDriver);
+        DriverFeature.getDriverManager().setCurrentDriver(commonTransformedDriver);
+
+        // TransformationDriver for Special Line Simulator
+        commonTransformedDriver = new TransformingDriver(new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special"));
```
Configure transformation explicitly.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/TransformingDriver.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class TransformingDriver implements Job2dDriver {
+    private final Job2dDriver driver;
+    private double scaleX;
+    private double scaleY;
+    private double rotationAngle;
+    private boolean isFlipped;
+    private int offsetX;
+    private int offsetY;
```
Try refactoring the proposed solution to a more OCP (open-closed principle) approach. Adding a new transformation is now possible by:
1. modifying this class - not 'closed' for modification,
2. adding a new class similar to TransformingDriver - would make an unclear division of responsibility between those classes.

kowallus reviewed (changes requested):
Go for more OCP design. Following SRP principle can be also a good idea.

Mirek0206 committed:
Approach changed to "Chain of Responsibility" to cover OCP rule.

Mirek0206 committed:
Transformations configured explicitly.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/drivers/DriverManager.java:
```diff
@@ -15,7 +16,7 @@ public class DriverManager {
      */
     public synchronized void setCurrentDriver(Job2dDriver driver) {
-        currentDriver = driver;
+        currentDriver = new TransformingDriver(driver);
```
Please do not burden DriverManager with TransformationDriver. It is not necessary. If you want to preserve your idea, you can use the observer approach. It is available in a recently merged PR #7

kowallus replied:
Question: Is it actually necessary to wrap the driver with TransformingDriver?

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -63,12 +65,30 @@ private static void setupDrivers(Application application) {
+        commonTransformedDriver.addModifier(new RotationModifier(0));
+        commonTransformedDriver.addModifier(new XAxisMovingModifier(0));
+        commonTransformedDriver.addModifier(new YAxisMovingModifier(0));
+        //commonTransformedDriver.addModifier(new VerticalFlipModifier());
+        //commonTransformedDriver.addModifier(new HorizontalFlipModifier());
```
Please remove unused code

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -63,12 +65,30 @@ private static void setupDrivers(Application application) {
         DriverFeature.addDriver("Precise logger driver", preciousLoggerDriver);
 
         DrawPanelController drawerController = DrawerFeature.getDrawerController();
-        Job2dDriver driver = new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic");
-        DriverFeature.addDriver("Line Simulator", driver);
-        DriverFeature.getDriverManager().setCurrentDriver(driver);
 
-        driver = new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special");
-        DriverFeature.addDriver("Special line Simulator", driver);
```
Do not remove existing drivers, add new ones

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -63,12 +65,30 @@ private static void setupDrivers(Application application) {
-        DriverFeature.addDriver("Special line Simulator", driver);
+        // TransformationDriver for Line Simulator
+        TransformingDriver commonTransformedDriver = new TransformingDriver(new LineDriverAdapter(drawerController, LineFactory.getBasicLine(), "basic"));
+        commonTransformedDriver.addModifier(new ScalingModifier(1.0, 1.0));
+        commonTransformedDriver.addModifier(new RotationModifier(0));
+        commonTransformedDriver.addModifier(new XAxisMovingModifier(0));
+        commonTransformedDriver.addModifier(new YAxisMovingModifier(0));
```
What is the point of those modifiers? They do not change anything. 
Propose a transformation (describe it shortly in a driver label)

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/driverTransofrmation/modifiers/VerticalFlipModifier.java:
```diff
@@ -0,0 +1,13 @@
+    }
+
+    @Override
+    public int[] modify(int x, int y) {
+        return new int[]{ x, (y * -1)};
+    }
```
this looks like a special case of ScalingModifier. Instead of VerticalFlipModifier (and other similar), please add a factory of ScalingModifier preset instances.

kowallus commented on src/main/java/edu/kis/powp/jobs2d/features/driverTransofrmation/modifiers/YAxisMovingModifier.java:
```diff
@@ -0,0 +1,16 @@
+
+import edu.kis.powp.jobs2d.features.driverTransofrmation.TransformationModifier;
+
+public class YAxisMovingModifier implements TransformationModifier {
+    private final int shiftY;
+
+    public YAxisMovingModifier(int shiftY) {
+        this.shiftY = shiftY;
+    }
+
+    @Override
+    public int[] modify(int x, int y) {
+        return new int[]{x, (y + shiftY)};
+    }
+}
```
OK, but  XAxisMovingModifier and  YAxisMovingModifier could be implemented as a single class.

kowallus reviewed (changes requested):
Much better, but there is some code cleaning to do, and provide better example drivers.

Mirek0206 committed:
Commented lines have been removed. Y and X axis modifier have been merged to the one class.

Mirek0206 committed:
Removed unused class.

Mirek0206 committed:
DriverManager - revoke changes.
TestJobs2dApp - restore basic drivers and use only needed modifiers.
TransformingDriver - added toString() method.

Mirek0206 committed:
Removed redundant files and added scaling modifier factory.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
-        DriverFeature.addDriver("Line Simulator", commonTransformedDriver);
+        TransformingDriver commonTransformedDriver = new TransformingDriver(driver);
+        commonTransformedDriver.addModifier(new ScalingModifier(0.25, 0.25));
+        DriverFeature.addDriver("Line Simulator  - transformed", commonTransformedDriver);
```
"transformed" - does not explain the expected effect. e.g. "1/4 scale"

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
+        TransformingDriver commonTransformedDriver = new TransformingDriver(driver);
+        commonTransformedDriver.addModifier(new ScalingModifier(0.25, 0.25));
+        DriverFeature.addDriver("Line Simulator  - transformed", commonTransformedDriver);
         DriverFeature.getDriverManager().setCurrentDriver(commonTransformedDriver);
```
do not setCurrentDriver here. Only one can be set at the same time. The default choice is in line 72.

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
-
-        DriverFeature.addDriver("Special Line Simulator", commonTransformedDriver);
+        commonTransformedDriver = new TransformingDriver(driver);
+        commonTransformedDriver.addModifier(new ScalingModifier(0.25, 0.25));
```
maybe test another transformation? or combination of several transformations!

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
-        DriverFeature.addDriver("Special Line Simulator", commonTransformedDriver);
+        commonTransformedDriver = new TransformingDriver(driver);
+        commonTransformedDriver.addModifier(new ScalingModifier(0.25, 0.25));
+        DriverFeature.addDriver("Special Line Simulator  - transformed", commonTransformedDriver);
```
"transformed" -> "..."?

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
+        commonTransformedDriver = new TransformingDriver(driver);
+        commonTransformedDriver.addModifier(new ScalingModifier(0.25, 0.25));
+        DriverFeature.addDriver("Special Line Simulator  - transformed", commonTransformedDriver);
+        DriverFeature.getDriverManager().setCurrentDriver(commonTransformedDriver);
```
remove

kowallus commented on src/test/java/edu/kis/powp/jobs2d/TestJobs2dApp.java:
```diff
@@ -66,28 +66,27 @@ private static void setupDrivers(Application application) {
+        // Driver for Special Line Simulator
+        driver = new LineDriverAdapter(drawerController, LineFactory.getSpecialLine(), "special");
+        DriverFeature.addDriver("Special line Simulator", driver);
+        DriverFeature.getDriverManager().setCurrentDriver(driver);
```
remove this line

kowallus reviewed (changes requested):
One more time ;). Minor issues

Mirek0206 committed:
Removed unnecessary usages of setCurrentDriver()

Mirek0206 committed:
Performed additional transformations and changed naming approach.

Mirek0206 committed:
Comment corrected.
